R/cnd_df.R

```diff
@@ -0,0 +1,321 @@
+#'
+#' @inheritParams pillar::ctl_new_rowid_pillar
+#' @importFrom pillar ctl_new_rowid_pillar
+#' @export
+ctl_new_rowid_pillar.cnd_df <- function(controller, x, width, ...) {
+
+  out <- NextMethod()
+  n_row <- nrow(x)
+  idx <- seq_len(n_row)
+  i <- sprintf("%d", idx)
+  i_width <- nchar(as.character(i))
+  i_max_width <- max(i_width)
+  max_width <- i_max_width + 2
+  ws <- strrep(" ", max_width - i_width - 1)
+  abb_lgl <- lgl_to_chr(attr(controller, "cnd")[idx])
+
+  row_ids <- paste0(i, ws, abb_lgl)
+  width <- max(nchar(as.character(row_ids)))
+  pillar::new_pillar(
+    list(
+      title = out$title,
+      type = out$type,
+      data = pillar::pillar_component(
+        pillar::new_pillar_shaft(list(row_ids = row_ids),
+                                 width = width,
+                                 class = "pillar_rif_shaft"
+        )
+      )
+    ),
+    width = width
+  )
+}
+
+#' Evaluate conditions
+#'
+#' @description
+#' [eval_conditions()] evaluates a set of conditions in the context of a
+#' data frame and an optional environment.
+#'
+#' The utility of this function is to provide an easy way to generate a logical
+#' vector of matching records from a set of logical conditions involving
+#' variables in a data frame (`dat`) and optionally in a supplementary
+#' environment (`.env`). The set of logical conditions are provided as
+#' expressions to be evaluated in the context of `dat` and `.env`.
+#'
+#' Variables are looked up in `dat`, then in `.env`, then in the calling
+#' function's environment, followed by its parent environments.
+#'
+#' @param dat A data frame
+#' @param ... A set of logical conditions, e.g. `y & z, x | z` (`x`, `y`, `z`
+#'   would have to exist either as columns in `dat` or in the environment
+#'   `.env`). If multiple expressions are included, they are combined with the
+#'   `&` operator.
+#' @param .na Return value to be used when the conditions evaluate to `NA`.
+#' @param .env An optional environment to look for variables involved in logical
+#'   expression passed in `...`. A data frame or a list can also be passed that
+#'   will be coerced to an environment internally.
+#'
+#' @returns A logical vector reflecting matching rows in `dat`.
+#'
+#' @examples
+#' # Create a sample data frame
+#' df <- data.frame(
+#'   x = c(1, 2, 3, 4, 5),
+#'   y = c(TRUE, FALSE, TRUE, FALSE, TRUE),
+#'   z = c("a", "b", "a", "b", "a")
+#' )
+#'
+#' # Simple condition on one column
+#' sdtm.oak:::eval_conditions(df, x > 2)
+#'
+#' # Combined conditions on multiple columns
+#' sdtm.oak:::eval_conditions(df, x > 2, y)
+#'
+#' # Using conditions with NA handling
+#' df_with_na <- data.frame(
+#'   x = c(1, 2, NA, 4, 5),
+#'   y = c(TRUE, FALSE, TRUE, FALSE, TRUE)
+#' )
+#' sdtm.oak:::eval_conditions(df_with_na, x > 2, .na = FALSE)
+#'
+#' # The environment where `eval_conditions()` is called is also inspected
+#' # when evaluating conditions in `...`.
+#' w <- 1
+#' sdtm.oak:::eval_conditions(df, x > w)
```

Comment:
> It will be helpful for the user to specify the dataframe and then specify the environment it resides in as two separate arguments.

The intention in having `.env` is not to indicate where the main data frame resides but to specify a namespace for looking for other variables that might be needed in the conditions supplied and that do not exist in the data frame `dat`. So you really should see `.env` as an object that provides a bag full of variables/vectors to be looked up for evaluation in the conditions given in `...` (if not part already of `dat`).

Also note that: `eval_conditions()` is currently an internal function, so it is not meant to be used by the user directly.

Reply:
ok, that makes it easier. We will just specify the .dat2 as the additional parameter.